### PR TITLE
Users/shreyas r msft/mocha test result parser refactor

### PR DESCRIPTION
**Pull request by ShreyasRmsft**

### CODE REVIEW

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs [hunk 1]
```diff
@@ -66,133 +70,119 @@ public MochaTestResultParser(ITestRunManager testRunManager, ITraceLogger logger
...
-                case MochaTestResultParserState.ExpectingStackTraces:
+                // State model for the mocha parser that defines the regexes to match against in each state
+                // Each state re-orders the regexes based on the frequency of expected matches
+                switch (this.currentState)
+                {
+                    // This state primarily looks for test results 
+                    // and transitions to the next one after a line of summary is encountered
+                    case MochaTestResultParserState.ExpectingTestResults:
+
+                        if (AttempMatch(this.expectingTestResults, testResultsLine))
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs]: >AttempMatch [](start = 28, length = 11)

Attempt*

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs [hunk 2]
```diff
@@ -66,133 +70,119 @@ public MochaTestResultParser(ITestRunManager testRunManager, ITraceLogger logger
...
+
+                // Rethrowing this so that the plugin is aware that the parser is erroring out
+                // Ideally this never should happen
+                throw;
             }
         }
 
         /// <summary>
-        /// Publishes the run and resets the parser by resetting the state context and current state
+        /// Attemps to match the line with each regex specified by the current state
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs]: >Attemps [](start = 12, length = 7)

Attempts*

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserStateContext.cs [hunk 3]
```diff
@@ -1,16 +1,71 @@
...
+
+        /// <summary>
+        /// The current line number of the input console log line
+        /// </summary>
+        public int CurrentLineNumber { get; set; }
+
+        /// <summary>
+        /// Used to reset the state context
+        /// </summary>
+        public override void Reset()
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserStateContext.cs]: >Reset [](start = 29, length = 5)

Send TestRun

File: src/Agent.Plugins/Parser/Node/Mocha/States/MochaParserStateBase.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Text.RegularExpressions;
+    using Agent.Plugins.TestResultParser.Loggers.Interfaces;
+    using Agent.Plugins.TestResultParser.Parser.Interfaces;
+    using Agent.Plugins.TestResultParser.Telemetry.Interfaces;
+    using Agent.Plugins.TestResultParser.TestResult.Models;
+
+    public abstract class MochaParserStateBase : ITestResultParserState
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/MochaParserStateBase.cs]: Comments

File: src/Agent.Plugins/Parser/Interfaces/ITestResultParserState.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Agent.Plugins.TestResultParser.Parser.Interfaces
+{
+    using System.Collections.Generic;
+    using Agent.Plugins.TestResultParser.Parser;
+
+    public delegate void ParserResetAndAttempPublish();
+
+    public interface ITestResultParserState
+    {
+        List<RegexActionPair> RegexesToMatch { get; }
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Interfaces/ITestResultParserState.cs]: >List [](start = 8, length = 4)

IEnumerable

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        public ExpectingTestResults(ParserResetAndAttempPublish parserResetAndAttempPublish)
+            : this(parserResetAndAttempPublish, TraceLogger.Instance, TelemetryDataCollector.Instance)
+        {
+
+        }
+
+        public ExpectingTestResults(ParserResetAndAttempPublish parserResetAndAttempPublish, ITraceLogger logger, ITelemetryDataCollector telemetryDataCollector)
+            : base(parserResetAndAttempPublish, logger, telemetryDataCollector)
+        {
+            RegexesToMatch = new List<RegexActionPair>
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs]: move out

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+            {
+                new RegexActionPair(MochaTestResultParserRegexes.PassedTestCase, PassedTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.FailedTestCase, FailedTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.PendingTestCase, PendingTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.PassedTestsSummary, PassedTestsSummaryMatched)
+            };
+
+            this.logger = logger;
+            this.telemetryDataCollector = telemetryDataCollector;
+            this.attemptPublishAndResetParser = parserResetAndAttempPublish;
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs]: redundant

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        }
+
+        private Enum PassedTestCaseMatched(Match match, TestResultParserStateContext stateContext)
+        {
+            var mochaStateContext = stateContext as MochaTestResultParserStateContext;
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs]: Can we avoid this?

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+            {
+                new RegexActionPair(MochaTestResultParserRegexes.PendingTestsSummary, PendingTestsSummaryMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.FailedTestsSummary, FailedTestsSummaryMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.PassedTestCase, PassedTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.FailedTestCase, FailedTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.PendingTestCase, PendingTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.PassedTestsSummary, PassedTestsSummaryMatched),
+            };
+
+            this.logger = logger;
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs]: redundant

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+            var testResult = new TestResult();
+
+            PrepareTestResult(testResult, TestOutcome.Passed, match);
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs]: Refactor new into PrepareTestResult

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+            return MochaTestResultParserState.ExpectingTestResults;
+        }
+
+        private Enum FailedTestCaseMatched(Match match, TestResultParserStateContext stateContext)
+        {
+            var mochaStateContext = stateContext as MochaTestResultParserStateContext;
+            var testResult = new TestResult();
+
+            // Handling parse errors is unnecessary
+            int.TryParse(match.Groups[RegexCaptureGroups.FailedTestCaseNumber].Value, out int testCaseNumber);
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs]: >TryParse [](start = 16, length = 8)

Parse

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+            this.logger.Info("MochaTestResultParser : ExpectingTestRunSummary : Transitioned to state ExpectingTestRunSummary.");
+            return MochaTestResultParserState.ExpectingTestRunSummary;
+        }
+
+        private Enum PendingTestsSummaryMatched(Match match, TestResultParserStateContext stateContext)
+        {
+            var mochaStateContext = stateContext as MochaTestResultParserStateContext;
+
+            this.logger.Info($"MochaTestResultParser : ExpectingTestRunSummary : Pending tests summary encountered at line {mochaStateContext.CurrentLineNumber}.");
+            mochaStateContext.LinesWithinWhichMatchIsExpected = 1;
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs]: ExpectedMatch

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+
+            // Handling parse errors is unnecessary
+            int.TryParse(match.Groups[RegexCaptureGroups.PendingTests].Value, out int totalPending);
+
+            mochaStateContext.TestRun.TestRunSummary.TotalSkipped = totalPending;
+
+            // If encountered skipped tests does not match summary fire telemtry
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs]: >telemtry [](start = 72, length = 8)

telemetry*

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+            mochaStateContext.StackTracesToSkipParsingPostSummary = totalFailed;
+
+            // If encountered failed tests does not match summary fire telemtry
+            if (mochaStateContext.TestRun.TestRunSummary.TotalFailed != mochaStateContext.TestRun.FailedTests.Count)
+            {
+                this.logger.Error($"MochaTestResultParser : ExpectingTestRunSummary : Failed tests count does not match failed summary" +
+                    $" at line {mochaStateContext.CurrentLineNumber}");
+                this.telemetryDataCollector.AddToCumulativeTelemtery(TelemetryConstants.EventArea,
+                    TelemetryConstants.PassedSummaryMismatch, new List<int> { mochaStateContext.TestRun.TestRunId }, true);
+            }
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs]: Avoid doing this here, move to common place.

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingStackTraces.cs [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+                this.logger.Error($"MochaTestResultParser : ExpectingStackTraces : Expecting stack trace with" +
+                    $" number {mochaStateContext.LastFailedTestCaseNumber + 1} but found {testCaseNumber} instead");
+                this.telemetryDataCollector.AddToCumulativeTelemtery(TelemetryConstants.EventArea, TelemetryConstants.UnexpectedFailedStackTraceNumber,
+                    new List<int> { mochaStateContext.TestRun.TestRunId }, true);
+
+                // If it was not 1 there's a good chance we read some random line as a failed test case hence consider it a
+                // as a match but do not consider it a valid stack trace
+                if (testCaseNumber != 1)
+                {
+                    return MochaTestResultParserState.ExpectingStackTraces;
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingStackTraces.cs]: >    return MochaTestResultParserState.ExpectingStackTraces; [](start = 16, length = 59)

take care during stack traces

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingStackTraces.cs [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+            // or incomplete stack trace data
+            // This check is safety check for when we try to parse stack trace contents, as of now it will always evaluate to true
+            if (mochaStateContext.StackTracesToSkipParsingPostSummary != 0)
+            {
+                this.logger.Error($"MochaTestResultParser : ExpectingStackTraces :  Expecting stack traces but found passed test case instead at line {mochaStateContext.CurrentLineNumber}.");
+                this.telemetryDataCollector.AddToCumulativeTelemtery(TelemetryConstants.EventArea, TelemetryConstants.ExpectingStackTracesButFoundPassedTest,
+                    new List<int> { mochaStateContext.TestRun.TestRunId }, true);
+            }
+
+            this.attemptPublishAndResetParser();
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingStackTraces.cs]: Move this is to the top

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingStackTraces.cs [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+                new RegexActionPair(MochaTestResultParserRegexes.FailedTestCase, FailedTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.PassedTestCase, PassedTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.PendingTestCase, PendingTestCaseMatched),
+                new RegexActionPair(MochaTestResultParserRegexes.PassedTestsSummary, PassedTestsSummaryMatched)
+            };
+        }
+
+        private Enum PassedTestCaseMatched(Match match, TestResultParserStateContext stateContext)
+        {
+            var testResult = new TestResult();
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingStackTraces.cs]: Move these down

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs [hunk 18]
```diff
@@ -31,11 +31,15 @@ public class MochaTestResultParser : ITestResultParser
         private MochaTestResultParserStateContext stateContext;
         private int currentTestRunId = 1;
 
-        private MochaTestResultParserState state;
+        private MochaTestResultParserState currentState;
         private ITraceLogger logger;
         private ITelemetryDataCollector telemetryDataCollector;
         private ITestRunManager testRunManager;
 
+        ITestResultParserState expectingTestResults;
```
Comment (cltshivash) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParser.cs]: >ITestResultParserState [](start = 8, length = 22)

nitpick: qualifier

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserRegexes.cs [hunk 19]
```diff
@@ -11,11 +11,13 @@ namespace Agent.Plugins.TestResultParser.Parser.Node.Mocha
     // TODO: Verify if tabs (/t) will come from the agent logs
     public class MochaTestResultParserRegexes
     {
-        public static Regex PassedTestCase { get; } = new Regex($"^(  )+((ΓêÜ)|✓) (((?<{RegexCaptureGroups.TestCaseName}>.*) \\((?<{RegexCaptureGroups.TestRunTime}>[0-9]+)(?<{RegexCaptureGroups.TestRunTimeUnit}>ms|s|m|h)\\)$)|(?<{RegexCaptureGroups.TestCaseName}>.*)$)", RegexOptions.ExplicitCapture);
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserRegexes.cs]: Comments for regex, good to an example

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs (hunk 8 above)
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestResults.cs]: Tried to think of a way. Only way is to have a wrapper class for the enum or mush all enums together. Both of them seem either equally or more inelegant than the current solution. Leaving this as is

File: src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs (hunk 14 above)
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Parser/Node/Mocha/States/ExpectingTestRunSummary.cs]: done

File: src/Agent.Plugins/Parser/Node/Mocha/States/MochaParserStateBase.cs (hunk 4 above)
Comment (ShreyasRmsft) [reply src/Agent.Plugins/Parser/Node/Mocha/States/MochaParserStateBase.cs]: done

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserStateContext.cs [hunk 20]
```diff
@@ -1,16 +1,61 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Agent.Plugins.TestResultParser.TestResult.Models;
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserStateContext.cs]: using inside the namespace

File: src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserStateContext.cs [hunk 21]
```diff
@@ -1,16 +1,61 @@
...
+    public class MochaTestResultParserStateContext : TestResultParserStateContext
     {
-        public int StackTracesToSkipParsingPostSummary { get; set; } = 0;
+        public MochaTestResultParserStateContext(TestRun testRun)
+        {
+            Initialize(testRun);
+        }
+
+        /// <summary>
+        /// Test run associted with the current iteration of the parser
```
Comment (singhsarab) [inline src/Agent.Plugins/Parser/Node/Mocha/MochaTestResultParserStateContext.cs]: Nit: associated*

**Pull request merged**